### DISCUSSION TITLE
[Section 5: Upgradable contract] what does __UUPSUpgradeable_init() actually do? what will happen if it is removed

### DISCUSSION THREAD

**Discussion by Vincent71399** [Cyfrin/foundry-full-course-cu#2613] [Q&A]
I tried to comment out this line __UUPSUpgradeable_init(), the test is still working

I also tried to read into the code, I did not find any logic behind it

can someone give me an example of not calling __UUPSUpgradeable_init() in initializer and cause error?

**Comment by EngrPips** [comment]
I have also looked into this and could not find the implementation. Let's look forward to @PatrickAlphaC  response.

**Reply by Vincent71399** [reply]
Thanks for the reply

**Reply by EngrPips** [reply]
You are welcome.

**Comment by yawarasuuna** [comment]
`__UUPSUpgradeable_init()` doesn't do anything, it's just a best practice that signals contracts as an UUPS Upgradeable, and improving readability

**Reply by EngrPips** [reply]
Interesting

**Reply by programmer-ke** [reply]
With the v5.6.1 of openzeppelin/contracts-upgradeable it is missing from the openzeppelin contracts and compiling fails if it is included, so I guess its not needed in newer the versions.

**Reply by EngrPips** [reply]
That makes sense.

**Comment by BhariGowda** [comment]
`__UUPSUpgradeable_init()` is an **initializer chain call** — it exists for safety and future-proofing, not current logic.

**Why it appears to do nothing:**
In the current OpenZeppelin implementation, `__UUPSUpgradeable_init()` has an empty body. That's why commenting it out doesn't break your tests — there's literally no logic in it right now.

**Why you should still call it:**

1. **Initializer chain convention**  OpenZeppelin's upgradeable contracts follow a pattern where every contract in the inheritance chain calls its own `__X_init()`. This ensures if OZ adds logic to it in a future version, your contract automatically inherits it without needing changes.

2. **`__UUPSUpgradeable_init_unchained()` separation**  OZ splits init logic into `_init()` (calls parent inits) and `_init_unchained()` (only own logic). Following this pattern keeps your contract compatible with the full OZ upgrade ecosystem.

3. **Auditor expectations** Security auditors flag missing `__X_init()` calls as a potential issue even if currently empty, because it signals the dev didn't follow the initialization pattern correctly.

**Concrete example where skipping causes problems:**
If OZ ever adds storage slot validation or ERC-1967 slot checks inside `__UUPSUpgradeable_init()`, contracts that skipped it would be vulnerable after upgrading their OZ dependency.

**Rule of thumb:** Always call `__X_init()` for every contract in your inheritance chain, even if empty today.
